pricegraph/src/orderbook.rs

```diff
@@ -118,22 +121,29 @@ impl Orderbook {
 
     /// Reduces the orderbook by matching all overlapping ring trades.
     pub fn reduce_overlapping_orders(mut self) -> ReducedOrderbook {
-        Subgraphs::new(self.projection.node_indices()).for_each(|token| loop {
-            match bellman_ford::search(&self.projection, token) {
-                Ok((_, predecessors)) => break predecessors,
-                Err(NegativeCycle(predecessors, node)) => {
-                    let path = path::find_cycle(&predecessors, node, None)
-                        .expect("negative cycle not found after being detected");
-
-                    self.fill_path(&path).unwrap_or_else(|| {
-                        panic!(
-                            "failed to fill path along detected negative cycle {}",
-                            format_path(&path),
-                        )
-                    });
-                }
+        let mut remaining_tokens: BTreeSet<_> = self.projection.node_indices().collect();
```

Comment:
I think this can be implemented with the `Subgraph` abstraction if instead of accepting a `ShortestPathGraph` it accepts a `Vec<NodeIndex>` as it did before and return `shortest_path_graph.connected_nodes()` from the closure.